docs/design/pnc.md

```diff
@@ -0,0 +1,192 @@
+  # this object allows additional parameters, you can put any metadata here
+  author: shadowman
+builds:
+  # all artifacts are grouped by builds to keep track of their sources
+  - build_id: '1234' # build id must be string
+    artifacts:
+      # list of artifacts to fetch, artifacts are fetched from PNC using their IDs
+      - id : '12345' # artifact id must be string
+        # the target can just be a filename or path+filename
+        target: test/rhba-common-7.10.0.redhat-00004.pom
+      - id: '12346'
+        target: prod/rhba-common-7.10.0.redhat-00004-dist.zip
+  - build_id: '1235'
+    artifacts:
+      - id: '12354'
+        target: test/client-patcher-7.10.0.redhat-00004.jar
+      - id: '12355'
+        target: prod/rhdm-7.10.0.redhat-00004-update.zip
+```
+
+These artifacts are fetched to `artifacts/<target>` path at the root of the repository.
+
+### PNC instance configuration
+Currently, the url of the PNC instance is configured as a part of OSBS config. 
+
+
+
+[atomic_reactor/plugins/fetch_maven_artifacts.py](https://github.com/containerbuildsystem/atomic-reactor/blob/master/atomic_reactor/plugins/fetch_maven_artifacts.py)
+1) OSBS loads the `fetch-artifacts-pnc.yaml`
+2) Information about individual artifacts is fetched from PNC API at `/artifacts/{id}` endpoint ([docs](https://github.com/project-ncl/pnc-api/blob/5b35ad6fe22769510f60b0025752dc6c954c9734/src/main/java/org/jboss/pnc/api/repositorydriver/dto/RepositoryArtifact.java))
+   - `publicUrl` key is used to get download url
+   - supported checksums are saved
+   - build ids are saved
+3) Download queue is formed containing above info for each artifact
+4) SBOM component info for each artifact is fetched from PNC @ `/artifacts/{id}`, `purl` key
+5) Queue is executed and each artifact is downloaded, checksum verified
+6) SBOM is compiled, including PNC components from step 4)
+
+
+### Missing features
+
+PNC SBOM components don't have the most complete purl, e.g.:
+`pkg:maven/org.kie.trustyai/explainability-service@999.0.0.managedsvc-redhat-00001?type=zip`
+
+According to [purl spec](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#maven), `classfier` and
+`type` parameters should be also defined, method is currently pending depending on if the artifacts are single components,
+or products that contain other components. Purl should also specify a `repository_url` parameter.
+
+
+
+## Design for the implementation in Cachi2
+
+### PNC config
+Existing `fetch-artifacts-pnc.yaml` file ([docs](https://osbs.readthedocs.io/en/osbs_ocp3/users.html?highlight=fetch#fetch-artifacts-pnc-yaml))
+will be used as configuration. It needs to be extended to also include PNC instance url.
+
+### Pre-fetch PNC dependencies
+Cachi2 will use the `artifacts` key specified in the `fetch-artifacts-pnc.yaml` file to fetch info about given artifacts
+from PNC (`/artifacts/{id}` endpoint), including url to download the artifact, checksums to verify integrity, and the purl.
```

Comment:
> checksums to verify integrity

Maybe we could add a H4 heading under this one to expand on how the checksum validation will work. I am assuming that:

- the checksums are always present in the API response, and if they're not, we'll report them as missing in the SBOM for that specific artifact
- if there's a checksum mismatch, we will fail the request straight away
- we can rely on a single algorithm type (since the API provides three different ones), probably sha-256?

(the first two assumptions derive from what Cachi2 is doing for other package managers)

Reply:
Since there are no user-defined checksums in fetch-artifacts-pnc.yaml, shouldn't all PNC SBOM components have the property `cachi2:missing_hash:in_file` regardless of whether the API response has them? Also, if the API response doesn't have them we should probably just fail since we cannot verify what was downloaded

Reply:
@brunoapimentel That is a correct assumption. However, based on what I saw in cachi2 code, you seem to be verifying all supported checksum algos, so I think for consistency, we should be doing all of them (even though sha-256 is clearly the strongest one)

@taylormadore I will contact prodsec to check, but I assume that the API response can be trusted.

Reply:
Taylor has a good point, that I overlooked in my first comment. We differentiate between server-provided and user-provided checksums, and the `cachi2:missing_hash:in_file` is only used for the latter.

I was wondering if we should expand the `fetch-artifacts-pnc.yaml` definition to allow the inclusion of checksums there, but then we would push it towards being more of a lockfile, which I'm not sure if it makes sense, considering the dependency we have on the PNC API to get more data about the artifact.

Another way to go is to simply consider that the PNC package manager will not verify user checksums, and properly document it.

Reply:
>  I will contact prodsec to check, but I assume that the API response can be trusted.
I think we should look at this from the perspective that it is a generic PNC support, in a sense that it extrapolates our immediate internal use-case. This means that any PNC instance could be used during the prefetching.

Reply:
>I think we should look at this from the perspective that it is a generic PNC support, in a sense that it extrapolates our immediate internal use-case. This means that any PNC instance could be used during the prefetching.

Okay, makes sense, will incorporate into the design.

@twaugh can you weigh in on the question of expanding the file with checksums for each artifact, from the perspective of users that will be migrating to konflux?

Reply:
>I think we should look at this from the perspective that it is a generic PNC support, in a sense that it extrapolates our immediate internal use-case. This means that any PNC instance could be used during the prefetching.

Prodsec view on this is that `cachi2:missing_hash:in_file` is reasonable to do. They did ask about if it would be possible to configure cachi2 to trust PNC as a source of checksums, off by default?